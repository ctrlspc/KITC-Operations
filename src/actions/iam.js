import database, { firebase, googleAuthProvider} from  '../firebase/firebase';

export const login = ({uid,email, displayName, photoURL }) => ({
  type: 'LOGIN',
  user: {uid, email, displayName, photoURL}
});

export const startLogin = () => {
  return (dispatch) => {
    return firebase.auth().signInWithPopup(googleAuthProvider);
  };
};

export const logout = () => ({
  type: 'LOGOUT'
});

export const startLogout = () => {
  return () => {
    return firebase.auth().signOut();
  };
};

export const setProfile = (profile) => ({
  type: 'SET_PROFILE',
  profile
});

export const startSetProfile = (user) => {
  return (dispatch) => {
    return database.ref(`profiles/${user.uid}`).once('value').then((snapshot) => {
      
      if(snapshot.val()){
        dispatch(setProfile({
          uid:snapshot.key,
          ...snapshot.val()
        }));
      }
    });
  };
};

export const setRoles = (roles) => ({
  type: 'SET_ROLES',
  roles
});

export const startSetRoles = (user) => {
  return(dispatch) => {
    return database.ref(`roles/${user.uid}`).once('value').then((snapshot) => {
      if(snapshot.val()) {
        dispatch(setRoles(Object.keys(snapshot.val())));
      }
    });
  }
};


export const startRegisterUser = (user) => {
  return (dispatch) => {

    console.log('Starting User Registration', user);
    const updatedUser = {...user,startedRegistration:true }
    return database.ref(`profiles/${user.uid}`).set(updatedUser).then (() => {
      return dispatch(startSetProfile(user)).then(() => {
        return dispatch(startSetRoles(user));
      });
    });
  };
};
