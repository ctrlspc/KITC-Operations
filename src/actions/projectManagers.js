import database from '../firebase/firebase';

export const setProjectManagers = (projectManagers) => ({
  type:'SET_PROJECT_MANAGERS',
  projectManagers
});

export const startSetProjectManagers = () => {
  return (dispatch) => {
    return database
    .ref('projectManagers')
    .once('value')
    .then((snapshot) => {
      const projectManagers = [];
      snapshot.forEach(snapshotChild => {
        projectManagers.push({
          id:snapshotChild.key,
          ...snapshotChild.val()
        });
      });
      dispatch(setProjectManagers(projectManagers));
    });
  };
};