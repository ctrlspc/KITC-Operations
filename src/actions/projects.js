import database from '../firebase/firebase';

export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const startSetProjects = () => {
  return (dispatch) => {
    return database.ref(`projects`)
      .once('value')
      .then((snapshot) => {
        const projects = [];

        snapshot.forEach(childsnapshot => {
          projects.push({
            id:childsnapshot.key,
            ...childsnapshot.val()
          });
        });        
        dispatch(setProjects(projects));
      });
  };
};

export const addProject = (project) => ({
  type: 'ADD_PROJECT',
  project
});

export const startAddProject = ({title, description, projectManager, projectType}) => {
 return (dispatch) => {
    const project = {title, description, projectManager, projectType};

    return database.ref(`projects`)
      .push(project)
      .then((ref) => {
        dispatch(addProject({
          id:ref.key,
          ...project
        }));

        return Promise.resolve(ref.key);
      });
  };
};

export const updateProject = (id, updates) => ({
  type: 'UPDATE_PROJECT',
  id, 
  updates
});

export const startUpdateProject = (id, updates) => {
  return (dispatch) => {
    return database.ref(`projects/${id}`)
      .update(updates)
      .then((ref) => {
        dispatch(updateProject(id, updates));
      });
    };
};
