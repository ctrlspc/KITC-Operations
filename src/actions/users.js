import database from '../firebase/firebase';

export const setUsers = (users) => ({
  type:'SET_USERS',
  users
});

export const startSetUsers = () => {
  return (dispatch) => {
    return database
    .ref('profiles')
    .once('value')
    .then((snapshot) => {
      const users = [];
      snapshot.forEach(snapshotChild => {
        users.push({
          uid:snapshotChild.key,
          ...snapshotChild.val()
        });
      });
      dispatch(setUsers(users));
    });
  };
};