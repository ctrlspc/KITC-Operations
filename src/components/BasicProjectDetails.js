import React from 'react';

export default ({title, description, projectManager, projectType}) => (
  <div>
    <h3>{title}</h3>
    <p>{description}</p>
    <p><span>Project Manager:</span> {projectManager.name}</p>
  </div>
);