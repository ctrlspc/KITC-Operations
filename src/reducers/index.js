import { combineReducers } from 'redux';

import iamReducer, * as fromIamReducer from './iam';
import projectsReducer from './projects';
import usersReducer from './users';

const reducer = combineReducers({
  iam: iamReducer,
  projects: projectsReducer,
  users: usersReducer
});

export const isAuthenticated = (state) => fromIamReducer.isAuthenticated(state.iam);
export const isAuthenticatedUser = (state) => fromIamReducer.isAuthenticatedUser(state.iam);
export const hasUserProfile = (state) => fromIamReducer.hasUserProfile(state.iam);

export const getActiveUsers = (state) => state.users //TODO: cross reference users who are active users.

export default reducer;