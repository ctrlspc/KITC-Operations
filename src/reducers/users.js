const projectManaReducerDefaultState = [];

export default (state = projectManaReducerDefaultState, action) => {
  switch (action.type){
    case 'SET_USERS':
      return action.users;
    default:
      return state;
  };
};