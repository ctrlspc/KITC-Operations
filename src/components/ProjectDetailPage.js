import React from 'react';
import { connect } from 'react-redux';
import BasicProjectDetails from './BasicProjectDetails';
import BasicProjectDetailsForm from './BasicProjectDetailsForm';
import { startUpdateProject } from '../actions/projects';
import { getActiveUsers } from '../reducers';
import { startSetUsers } from '../actions/users';

export class ProjectDetailPage extends React.Component {

  constructor(props) {
    super(props);
    this.props.getUsers();
    this.state = {
      editBasicProjectDetails: !!props.editBasicProjectDetails
    };
  }

  onEditBasicProjectDetails = () => {
    this.setState(() => ({editBasicProjectDetails: true}));
  };

  onFinishedEditingBasicProjectDetails = (updates) => {
    this.props.updateProject(this.props.project.id, updates).then(() => {
      this.setState(() => ({editBasicProjectDetails: false}));
    });
  }

  render (){
    return (
      <div className="content-container">
        <div className="box-section">
          <div className="box-section__header">
            <h1>Basic Project Details</h1>
            {!this.state.editBasicProjectDetails && 
              <button className="button" onClick={this.onEditBasicProjectDetails}>Edit</button>
            }
          </div>
          <div className="box-section__body">
            {
              this.state.editBasicProjectDetails ? 
              (
                <BasicProjectDetailsForm 
                  onSubmit={this.onFinishedEditingBasicProjectDetails} 
                  projectManagers={this.props.users}
                  project={this.props.project}
                />
              ) : (
                <BasicProjectDetails {...this.props.project} onEditClick={this.onEditBasicProjectDetails}/>
              )
            }
          </div>
        </div>
      </div>
    );
  };
};

export const mapStateToProps = (state, props) => ({
  project: state.projects.find((project) => project.id === props.match.params.id),
  users: getActiveUsers(state)
});

export const mapDispatchToProps = (dispatch) => ({
  getUsers: () => dispatch(startSetUsers()),
  updateProject: (id, project) => dispatch(startUpdateProject(id, project))
})

export default connect(mapStateToProps, mapDispatchToProps)(ProjectDetailPage);