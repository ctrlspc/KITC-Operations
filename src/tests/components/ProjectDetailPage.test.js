import React from 'react';
import { shallow } from 'enzyme';
import { ProjectDetailPage, mapStateToProps, mapDispatchToProps } from '../../components/ProjectDetailPage';

import projects from '../fixtures/projects';
import users from '../fixtures/users';

import { startUpdateProject } from '../../actions/projects';
import { getActiveUsers } from '../../reducers';
jest.mock('../../actions/projects');
jest.mock('../../reducers');

let getUsers;

beforeEach(() => {
  getUsers = jest.fn();
});

test('should default to be in non-edit mode and render ProjectDetailPage component correctly', () => {
  const wrapper = shallow(
    <ProjectDetailPage project={projects[0]} getUsers={getUsers}/>
  );
  expect(wrapper).toMatchSnapshot();
});

test('show edit detail views', () => {
  const wrapper = shallow(
    <ProjectDetailPage project={projects[0]} editBasicProjectDetails={true} getUsers={getUsers}/>
  );
  expect(wrapper).toMatchSnapshot();
});

test('should toggle to edit mode when the edit button is clicked', () => {
  const wrapper = shallow(
    <ProjectDetailPage project={projects[0]} editBasicProjectDetails={false} getUsers={getUsers}/>
  );
  wrapper.find('button').simulate('click');
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {

  const updateProject = jest.fn(() => Promise.resolve({}));
  const project = projects[0];
  const history = { push: jest.fn() };
  const wrapper = shallow(
    <ProjectDetailPage 
      updateProject={updateProject} 
      history={history}
      project={project}
      editBasicProjectDetails={true}
      getUsers={getUsers}
    />
  );
  wrapper.find('BasicProjectDetailsForm').prop('onSubmit')(projects[0]);
  expect(updateProject).toHaveBeenLastCalledWith(projects[0].id, projects[0]);
  expect(wrapper.state.editBasicProjectDetails).toBeFalsy();
});

test('should map state to props correctly', () => {
  const state = {
    projects: projects
  };

  getActiveUsers.mockReturnValue(users);

  const props = {
    match: {params: {id: '1'}}
  };

  const map = mapStateToProps(state, props);
  expect(map.project).toEqual(projects[0]);
  expect(map.users).toEqual(users);
  expect(getActiveUsers).toHaveBeenCalledWith(state);

});

test('should correctly map dispatch to props', () => {
  const dispatch = jest.fn();
  const props = mapDispatchToProps(dispatch);
  startUpdateProject.mockReturnValue({test:true});
  props.updateProject(1, {uid:'123'});
  expect(startUpdateProject.mock.calls[0][0]).toEqual(1);
  expect(startUpdateProject.mock.calls[0][1]).toEqual({uid:'123'});
  expect(dispatch.mock.calls[0][0]).toEqual({test:true});
});