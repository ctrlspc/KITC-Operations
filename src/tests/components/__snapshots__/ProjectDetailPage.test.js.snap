// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should default to be in non-edit mode and render ProjectDetailPage component correctly 1`] = `
<div
  className="content-container"
>
  <div
    className="box-section"
  >
    <div
      className="box-section__header"
    >
      <h1>
        Basic Project Details
      </h1>
      <button
        className="button"
        onClick={[Function]}
      >
        Edit
      </button>
    </div>
    <div
      className="box-section__body"
    >
      <Component
        description="testing some great stuff"
        id="1"
        onEditClick={[Function]}
        projectManager={
          Object {
            "id": "1",
            "name": "Mr Test",
          }
        }
        projectType="int"
        title="first test project"
      />
    </div>
  </div>
</div>
`;

exports[`show edit detail views 1`] = `
<div
  className="content-container"
>
  <div
    className="box-section"
  >
    <div
      className="box-section__header"
    >
      <h1>
        Basic Project Details
      </h1>
    </div>
    <div
      className="box-section__body"
    >
      <BasicProjectDetailsForm
        onSubmit={[Function]}
        project={
          Object {
            "description": "testing some great stuff",
            "id": "1",
            "projectManager": Object {
              "id": "1",
              "name": "Mr Test",
            },
            "projectType": "int",
            "title": "first test project",
          }
        }
      />
    </div>
  </div>
</div>
`;
